MCprep_addon/commonmcobj_parser.py

```diff
@@ -0,0 +1,181 @@
+
+    header_lines list[str]: 
+        list of strings representing each line of the header.
+
+    returns:
+        CommonMCOBJ object
+    """
+
+    # Split at the colon and clean up formatting
+    def clean_and_extract(line: str) -> Tuple[str, str]:
+        split = line.split(':', 1)
+        pos = 0
+        for i,x in enumerate(split[0]):
+            if x.isalpha():
+                pos = i                      
+                break
+        return (split[0][pos:], split[1].strip())
+    
+    # Basic values
+    header = CommonMCOBJ(
+        version=0,
+        exporter="NULL",
+        world_name="NULL",
+        world_path="NULL",
+        export_bounds_min=(0, 0, 0),
+        export_bounds_max=(0, 0, 0),
+        export_offset=(0, 0, 0),
+        block_scale=0,
+        block_origin_offset=(0, 0, 0),
+        z_up=False,
+        texture_type=CommonMCOBJTextureType.ATLAS,
+        has_split_blocks=False
+    )
+    
+    # Keys whose values do not need extra processing
+    NO_VALUE_PARSE = [
+        "exporter", 
+        "world_name", 
+        "world_path",
+    ]
+
+    # Keys whose values are tuples
+    TUPLE_PARSE_INT = [
+        "export_bounds_min",
+        "export_bounds_max",
+    ]
+
+    TUPLE_PARSE_FLOAT = [
+        "export_offset",
+        "block_origin_offset"
+    ]
+    
+    # Keys whose values are booleans
+    BOOLEAN_PARSE = [
+        "z_up",
+        "has_split_blocks"
+    ]
+    
+    # Although CommonMCOBJ states that 
+    # order does matter in the header, 
+    # future versions may change the order
+    # of some values, so it's best to 
+    # use a non-order specific parser
+    for line in header_lines:
+        if ":" not in line:
+            continue 
+        key, value = clean_and_extract(line)
+
+        if key == "version":
+            header.version = int(value)
+
+        elif key == "block_scale":
+            header.block_scale = float(value)
+
+        elif key == "texture_type":
+            header.texture_type = CommonMCOBJTextureType[value]
+
+        # All of these are parsed the same, with 
+        # no parsing need to value
+        elif key in NO_VALUE_PARSE:
+            setattr(header, key, value)
+
+        # All of these are parsed the same, with 
+        # parsing the value to a tuple
+        elif key in TUPLE_PARSE_INT:
+            setattr(header, key, tuple(map(int, value[1:-1].split(', '))))
+
+        elif key in TUPLE_PARSE_FLOAT:
+            setattr(header, key, tuple(map(float, value[1:-1].split(', '))))
+
+        elif key in BOOLEAN_PARSE:
+            setattr(header, key, value == "true")
+
+    return header
+
+def parse_header(f: TextIO) -> Optional[CommonMCOBJ]:
+    """
+    Parses a file and returns a CommonMCOBJ object if 
+    the header exists.
+    
+    f: TextIO
+        File object
+    
+    Returns:
+        - CommonMCOBJ object if header exists
+        - None otherwise
+    """
+    lines = f.readlines()
```

Comment:
This is pretty memory intensive and almost always will load in more data than we need, since the header is at the top. This would load the entire obj into python memory (doublely bloated) while using an iterator would only have individual lines at a time up until we exit out.

Let's do a with-clause iterator where we read one line at a time, stopping after we reach a the `# COMMON_MC_OBJ_END` or maybe after reading 10 lines with no comments (`#`).

With each line, we can append to a list which we later .join() into a string (if relevant).

Reply:
Resolved in [`ab2e163`](https://github.com/Moo-Ack-Productions/MCprep/pull/555/commits/ab2e1635fb5d5752e98971dd833ccdb00ff54014)